diff --git a/libgralloc/framebuffer.cpp b/libgralloc/framebuffer.cpp
index 517b92c..5e17556 100644
--- a/libgralloc/framebuffer.cpp
+++ b/libgralloc/framebuffer.cpp
@@ -303,6 +303,16 @@ int mapFrameBufferLocked(struct private_module_t* module)
         ALOGW("page flipping not supported (yres_virtual=%d, requested=%d)",
               info.yres_virtual, info.yres*2);
     }
+   if (property_get("debug.gr.pageflip", property, NULL) > 0) {
+        int num = atoi(property);
+        if (num > 0) {
+            flags = PAGE_FLIP;
+            ALOGW("Page Flipping force enabled");
+        } else {
+            flags = 0;
+            ALOGW("Page Flipping force disabled");
+        }
+    }
 
     if (ioctl(fd, FBIOGET_VSCREENINFO, &info) == -1)
         return -errno;
